internal/proxy/reverse_proxy.go

```diff
@@ -0,0 +1,212 @@
+package proxy
+
+import (
+	"crypto/tls"
+	"fmt"
+	"net"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	log "github.com/buzzfeed/sso/internal/pkg/logging"
+)
+
+// upstreamTransport is used to ensure that upstreams cannot override the
+// security headers applied by sso_proxy. Also includes functionality to rotate
+// http.Transport objects to ensure tcp connection reset deadlines are not exceeded
+type upstreamTransport struct {
+	mux sync.Mutex
+
+	deadAfter     time.Time
+	resetDeadline time.Duration
+
+	transport          *http.Transport
+	insecureSkipVerify bool
+}
+
+// RoundTrip round trips the request and deletes security headers before returning the response.
+func (t *upstreamTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	resp, err := t.getTransport().RoundTrip(req)
+	if err != nil {
+		log.Error(err, "error in upstreamTransport RoundTrip")
+		return nil, err
+	}
+	return resp, err
+}
+
+// getTransport gets either a cached http transport or allocates a new one
+// We proactively rotate tcp connects by re-cycling the http.Transport every
+// resetDeadline period.
+func (t *upstreamTransport) getTransport() *http.Transport {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+
+	if t.transport == nil || time.Now().After(t.deadAfter) {
+		t.deadAfter = time.Now().Add(t.resetDeadline)
+		t.transport = &http.Transport{
+			Proxy: http.ProxyFromEnvironment,
+			DialContext: (&net.Dialer{
+				Timeout:   30 * time.Second,
+				KeepAlive: 30 * time.Second,
+				DualStack: true,
+			}).DialContext,
+			MaxIdleConns:          100,
+			IdleConnTimeout:       90 * time.Second,
+			TLSHandshakeTimeout:   10 * time.Second,
+			TLSClientConfig:       &tls.Config{InsecureSkipVerify: t.insecureSkipVerify},
+			ExpectContinueTimeout: 1 * time.Second,
+		}
+	}
+
+	return t.transport
+}
+
+func NewUpstreamReverseProxy(config *UpstreamConfig, signer *RequestSigner) (http.Handler, error) {
```

Comment:
are we linting in our CI?

Reply:
Yes, https://github.com/buzzfeed/sso/blob/master/scripts/test#L28

Is there something here that would break linting?